go/src/koding/kites/kloud/kloud/kloud.go

```diff
@@ -413,16 +409,46 @@ func (k *Kloud) HandleFunc(pattern string, f kite.HandlerFunc) *kite.Method {
 
 // Close closes the underlying connections.
 func (k *Kloud) Close() error {
+	k.Kite.Close()
+
+	var merr *multierror.Error
 	if k.metricsProxy != nil {
-		return k.metricsProxy.Close()
+		if err := k.metricsProxy.Close(); err != nil {
+			merr = multierror.Append(merr, err)
+		}
 	}
 
-	k.Kite.Close()
-	k.Stats.Close()
+	if err := k.Stats.Close(); err != nil {
+		merr = multierror.Append(merr, err)
+	}
+
+	defer k.closeOnce.Do(func() {
+		close(k.closeChan)
+	})
 
+	return merr.ErrorOrNil()
+}
+
+// Wait waits for Kloud to exit
+func (k *Kloud) Wait() error {
+	<-k.closeChan // wait for exit
 	return nil
 }
 
+func (k *Kloud) handleSignals() {
+	go func() {
+		signalCh := make(chan os.Signal, 1)
+		signal.Notify(signalCh)
+
+		s := <-signalCh
+		signal.Stop(signalCh)
+		switch s {
+		case syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL:
+			k.Kite.Log.Info("%s signal received, closing kloud", s)
+			k.Close()
+		}
```

Comment:
Usually second signal forcefully closes a process, e.g. sending CTRL+C twice. With for-loop it won't never stop, unless -9 is sent.